learnersync.go

```diff
@@ -207,7 +208,15 @@ func (s *Sync) postJSON(endpoint, data string) error {
 }
 
 func (s *Sync) PostPing() error {
-	return s.postJSON("pings", "")
+	body, err := json.Marshal(
+		struct {
+			InHostedEnv bool `json:"sent_from_hosted_environment"`
```

Comment:
The test looks good so 👍🏻

Buuuuuuuuuuuuut this feels like adding special-purpose complexity that ought to be contained within _train_.

Assuming the feature is "lights up cloud icon to show that hosted learner environment is working", I would do it by making _train_ generate a second session_id, then starting the HLE with that session_id. When _train_ receives that session_id back through an HTTP ping, it knows the HLE is working. That would mean no change to learnersync, and a shorter testing path for the feature.

For other coach-assisted debugging you could consider standard HTTP headers e.g.
* a [`User-Agent` header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/User-Agent) that shares the `Version` constant, 
* a [`From` header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/From) that shares `git config --get user.email` 🕵🏻

Reply:
In the HLE we could just add “-foo” when we write the attendance id file

Reply:
I take your point. But I think that will be fiddlier to implement in the short term, as it is implicitly quite tightly coupled with learner setup instructions.

Right now we're in a slightly awkward middle ground, where we have learner instructions that are designed for local setup, with learners provided with the attendance id that works there. Currently, if somebody follows those instructions exactly on the hosted environment, then things will still work fine. 

If we have a different attendance id for the hosted environment, then we'd need to do more work on all the curriculum setups & instructions to be sure that learners aren't inadvertently setting their 'local attendance id' on the hosted environment, which would make the data we collect about where they are doing their exercises incorrect.

I think that the coupling with curriculum means that changing the sync is a significantly smaller change right now, and allows this work to progress. However, I think we may well want to shift to something more like what you're suggesting when we are more explicitly differentiating between the setups for local and hosted environments.

Does that make sense?

Reply:
> For other coach-assisted debugging you could consider standard HTTP headers e.g.
>
> a [User-Agent header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/User-Agent) that shares the Version constant,
> a [From header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/From) that shares git config --get user.email 🕵🏻

Do you think this would be useful for external coaches? I think they could well be useful for us debugging, but I thought that coaches wouldn't be able to do much to solve problems live with those fields.

Reply:
> I take your point. But I think that will be fiddlier to implement in the short term, as it is implicitly quite tightly coupled with learner setup instructions.
> 
> Right now we're in a slightly awkward middle ground, where we have learner instructions that are designed for local setup, with learners provided with the attendance id that works there. Currently, if somebody follows those instructions exactly on the hosted environment, then things will still work fine.
> 
> If we have a different attendance id for the hosted environment, then we'd need to do more work on all the curriculum setups & instructions to be sure that learners aren't inadvertently setting their 'local attendance id' on the hosted environment, which would make the data we collect about where they are doing their exercises incorrect.

Sure, you absolutely don't want to make manual setup instructions any more complicated. But I don't understand why a learner would be setting up the sync manually in the hosted environment? They shouldn't even be aware of the sync client there.

Reply:
Right - that's the goal, but to get there we need to make changes to all the curricula instructions / emails going out to learners / train UX to avoid this, and make sure that learners are clear about what they need to do, where.

We've deliberately gone for a v1 which requires minimal changes to all of those things, so we can get things functional and start testing it quicker. This doesn't feel _too_ awful to me in terms of where responsibilities lie, but we'll expect to move towards fully background sync when that's important enough (e.g. probably when we need the HLE for beginners who struggle with setup, rather than for Docker learners).

I'm also not the most keen on encoding extra information in the attendance_id instead of passing extra parameters. It feels a little like a negative merging of concerns to me. I'm probably being a bit OTT there though.